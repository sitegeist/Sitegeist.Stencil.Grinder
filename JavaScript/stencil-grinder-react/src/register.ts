import { ComponentType } from "react";

/**
 * This file is part of the Sitegeist.Stencil.Grinder package
 *
 * (c) 2019
 * Wilhelm Behncke <behncke@sitegeist.de>
 *
 * This package is Open Source Software. For the full copyright and license
 * information, please view the LICENSE file which was distributed with this
 * source code.
 */

/**
 * Central registry for all components known to Grinder
 */
const registry: Map<string, ComponentType<any>> = new Map();

/**
 * @private
 * @param string key
 * @return ComponentType<any>
 */
export function get(key: string): ComponentType<any> {
	if (!registry.has(key)) {
		throw new Error(`[Grinder]: No component has been registered for key "${key}".`);
	}

	return registry.get(key) as ComponentType<any>;
}

/**
 * @public
 * @param string key
 * @param ComponentType implementation?
 * @return ComponentType<any> | (implementation: ComponentType<any>) => ComponentType<any>
 */
export default function register(key: string, implementation?: ComponentType<any>) {
	if (implementation) {
		registry.set(key, implementation);
		return implementation;
	}

	return function registerDecorator(implementation: ComponentType<any>) {
		registry.set(key, implementation);
		return implementation;
	}
}
