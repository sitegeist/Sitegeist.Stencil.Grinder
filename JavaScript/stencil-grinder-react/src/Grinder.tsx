import React from "react";
import { GrinderResponse } from "@sitegeist/stencil-grinder-shared";

import { get } from "./register";

/**
 * This file is part of the Sitegeist.Stencil.Grinder package
 *
 * (c) 2019
 * Wilhelm Behncke <behncke@sitegeist.de>
 *
 * This package is Open Source Software. For the full copyright and license
 * information, please view the LICENSE file which was distributed with this
 * source code.
 */

type GrinderData = GrinderResponse | { key: any, [key: string]: any } | object | number | string | null;

interface GrinderProps {
	data: GrinderData | GrinderData[]
}

/**
 * @public
 * @param GrinderProps props
 * @return JSX.Element[] | undefined
 */
export default function Grinder(props: GrinderProps): JSX.Element | null {
	if (Array.isArray(props.data)) {
		return (
			<>
				{props.data.map((item, index) => {
					if (item !== null && typeof item === 'object') {
						return (
							<Grinder key={'key' in item ? item.key : index} data={item} />
						);
					}

					return item;
				})}
			</>
		);
	}

	if (typeof props.data === 'object' && props.data !== null && 'type' in props.data) {
		switch (props.data.type) {
			case 'Sitegeist.Stencil.Grinder/v1/INDEX':
				throw new Error(`[Grinder]: Cannot render index. The index is not supposed to be rendered.`);

			case 'Sitegeist.Stencil.Grinder/v1/COMMAND':
				throw new Error(`[Grinder]: Cannot render command "${props.data.payload.directive}". Please make sure to handle commands before rendering.`);

			case 'Sitegeist.Stencil.Grinder/v1/DOCUMENT':
				return (
					<Grinder data={props.data.payload}/>
				);

			case 'Sitegeist.Stencil.Grinder/v1/NODE': {
				const Component = get(props.data.payload.component);

				if (Component) {
					return (<Component {...props.data.payload.props} />);
				}

				throw new Error(`[Grinder]: Could not find component: ${props.data.payload.component}`);
			}

			default: break;
		}
	}

	if (typeof props.data === 'string') {
		return (<>{props.data}</>);
	}

	if (typeof props.data === 'number') {
		return (<>props.data</>);
	}

	if (props.data === null) {
		return null;
	}

	throw new Error(`[Grinder]: Could not unravel data structure: ${JSON.stringify(props.data)}.`);
}
