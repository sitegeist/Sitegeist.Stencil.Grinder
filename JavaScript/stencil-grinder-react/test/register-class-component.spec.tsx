import React, { Component } from "react";
import renderer from "react-test-renderer";
import register, { get } from "../src/register";
import Grinder from "../src/Grinder";

describe(`Register class component`, () => {
	it(`should be able to register a class component`, () => {
		class MyClassComponent extends Component {
			render() {
				return (
					<div>Test</div>
				)
			}
		}

		register('MyComponent', MyClassComponent);

		expect(get('MyComponent')).not.toBe(null);
		expect(get('MyComponent')).toEqual(MyClassComponent);
	});

	it(`should render a registered class component correctly`, () => {
		class MyClassComponent extends Component {
			render() {
				return (
					<div>Test</div>
				)
			}
		}

		register('MyComponent', MyClassComponent);

		const rendered = renderer.create(
			<Grinder
				data={{
					type: 'Sitegeist.Stencil.Grinder/v1/NODE',
					payload: {
						component: 'MyComponent',
						props: {}
					}
				}}
			/>
		);

		expect(rendered.toJSON()).toMatchSnapshot();
	});
});
