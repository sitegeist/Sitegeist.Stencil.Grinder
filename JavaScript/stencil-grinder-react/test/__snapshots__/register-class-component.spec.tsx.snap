// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register class component should render a registered class component correctly 1`] = `
<div>
  Test
</div>
`;
